File: Graphs/Dijkstra/Dijkstra.java
package Graphs.Dijkstra;

import java.util.HashMap;
import java.util.Map;

public class Dijkstra {
    private static Map<String, Map<String, Integer>> graph = new HashMap<>();

    static {
        graph.put("start", new HashMap<>());
        graph.get("start").put("a", 6);
        graph.get("start").put("b", 2);

        graph.put("a", new HashMap<>());
        graph.get("a").put("end", 1);
        
        graph.put("b", new HashMap<>());
        graph.get("b").put("a", 3);
        graph.get("b").put("end", 5);

        graph.put("end", new HashMap<>());
    }

    private static final int INFINITY = Integer.MAX_VALUE;

    private static Map<String, Integer> costs = new HashMap<>();

    static {
        costs.put("a", 6);
        costs.put("b", 2);
        costs.put("end", INFINITY);
    }

    private static Map<String, String> parents = new HashMap<>();

    static {
        parents.put("a", "start");
        parents.put("b", "start");
        parents.put("end", null);
    }

    private static Map<String, Boolean> processed = new HashMap<>();

    public static void main(String[] args) {
        String node = findLowestCostNode(costs);

        while (node != null) {
            int cost = costs.get(node);
            Map<String, Integer> neighbors = graph.get(node);

            for (String neighbor : neighbors.keySet()) {
                int newCost = cost + neighbors.get(neighbor);
                if (costs.get(neighbor) > newCost) {
                    costs.put(neighbor, newCost);
                    parents.put(neighbor, node);
                }
            }

            processed.put(node, true);
            node = findLowestCostNode(costs);
        }

        System.out.println("Path: " + getPath());
    }

    private static String findLowestCostNode(Map<String, Integer> costs) {
        int lowestCost = INFINITY;
        String lowestCostNode = null;

        for (Map.Entry<String, Integer> entry : costs.entrySet()) {
            String node = entry.getKey();
            int cost = entry.getValue();
            if (cost < lowestCost && !processed.containsKey(node)) {
                lowestCost = cost;
                lowestCostNode = node;
            }
        }

        return lowestCostNode;
    }

    private static String getPath() {
        StringBuilder path = new StringBuilder("end");
        String parent = parents.get("end");

        while (parent != null) {
            path.insert(0, parent + " -> ");
            parent = parents.get(parent);
        }

        return path.toString();
    }
}

File: Graphs/Dijkstra/Dijkstra.py
graph = {
    "start": {"a": 6, "b": 2},
    "a": {"end": 1},
    "b": {"a": 3, "end": 5},
    "end": {}
}

INFINITY = float("inf")

costs = {
    "a": 6,
    "b": 2,
    "end": INFINITY
}

parents = {
    "a": "start",
    "b": "start",
    "end": None
}

processed = set()

def find_lowest_cost_node(costs):
    lowest_cost = INFINITY
    lowest_cost_node = None
    for node in costs:
        cost = costs[node]
        if cost < lowest_cost and node not in processed:
            lowest_cost = cost
            lowest_cost_node = node
    return lowest_cost_node

def get_path():
    path = ["end"]
    parent = parents["end"]
    while parent is not None:
        path.insert(0, parent)
        parent = parents[parent]
    return " -> ".join(path)

def main():
    node = find_lowest_cost_node(costs)
    while node is not None:
        cost = costs[node]
        neighbors = graph[node]
        for neighbor in neighbors:
            new_cost = cost + neighbors[neighbor]
            if costs.get(neighbor, INFINITY) > new_cost:
                costs[neighbor] = new_cost
                parents[neighbor] = node
        processed.add(node)
        node = find_lowest_cost_node(costs)
    
    print("Path:", get_path())

if __name__ == "__main__":
    main()
